src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,137 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (isEmpty()) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            if (size == 0) {
+                add(value);
+            } else {
+                Node<T> nextNode = head;
+                Node<T> newNode = new Node<>(null, value, nextNode);
+                head = newNode;
+                nextNode.prev = newNode;
+                size++;
+            }
+        } else if (index == size) {
+            add(value);
+        } else {
+            Node<T> prevNode = getNodeByIndex(index - 1);
+            Node<T> nextNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(prevNode, value, nextNode);
+            prevNode.next = newNode;
+            nextNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = getNodeByIndex(index).value;
+        if (size == 1) {
+            tail = null;
+            head = null;
+        } else if (index == 0) {
+            getNodeByIndex(index + 1).prev = null;
+            head = getNodeByIndex(index + 1);
+        } else if (index == size - 1) {
+            getNodeByIndex(index - 1).next = null;
+            tail = getNodeByIndex(index - 1);
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            Node<T> next = getNodeByIndex(index + 1);
+            prev.next = next;
+            next.prev = prev;
+        }
+        size--;
```

Comment:
try to simplify this code, you don't need to handle all these cases separately, You only need to call getNodeByIndex once